### ISSUE TITLE
Wrong loss calculation?

### ISSUE THREAD

**Issue by heshameraqi** [mvpcom/carlaILTrainer#19]
`tf.reduce_mean(means)` doesn't take the axis argument, so it will produce a single number that will be multiplied with the mask. So I think the line:
`contLoss = tf.reduce_sum(tf.multiply(tf.reduce_mean(means), mask))`
should be changed to:
`contLoss = tf.reduce_sum(tf.multiply(tf.transpose(tf.reduce_mean(means, 2)), mask))`

**Comment by heshameraqi** [comment]
Guys this is very important please to investigate and easy to have a look at, because how come do you achieve these good reported results using wrong train loss calculation?

**Comment by mvpcom** [comment]
@heshameraqi Hi, although both codes are here in the same repo for archive, AutoTableGenerator is different from the training code. You can use "AutoTableGenerator" after any experiment and with any code. Unfortunately, I wasn't able to find a proper time to continue to fix all the ToDo list. As you can see in the ToDo list, loss function needs to be fixed because the current version is both slow and inaccurate. Sorry for the misunderstanding. I don't speak about "train.py" which developed by Ivan Chernuha(@merryHunter ). 

I will publish the final code that I utilized for this purpose soon in this [repository](https://github.com/mvpcom/dodgeLearning). 

In the meanwhile, I suggest you use the following repositories to retrain the model more accurately and faster: [CoilTrain](https://github.com/felipecode/coiltraine) and [Imitation Learning](https://github.com/merantix/imitation-learning) by @markus-hinsche

Good Luck

**Comment by mvpcom** [comment]
@heshameraqi Please also consider Tenserflow is changing, and for many functions, the different behavior is possible. I mean in different versions, some features were altered.

**Comment by heshameraqi** [comment]
@mvpcom Thanks a lot for your reply. I understand AutoTableGenerator is an independent component and actually it's a nice contribution.

I've checked before raising the issue that `tf.reduce_mean` behavior did not change with TF versions.

A question please, for the results you show in the readme file, which trainer code is used to produce it?
Thanks a lot in advance.

**Comment by mvpcom** [comment]
@heshameraqi I didn't publish that code yet but I will soon [here](https://github.com/mvpcom/dodgeLearning).
